packages/core/src/stateUtils.ts

```diff
@@ -1665,7 +1665,7 @@ function stopStep(
   for (const stateNode of nextState.configuration.sort(
     (a, b) => b.order - a.order
   )) {
-    actions.push(...stateNode.definition.exit);
```

Comment:
This was quite a 🦋 effect. 

1. `definition` is a getter: https://github.com/statelyai/xstate/blob/68b24d7c4ce7fa8920edab0a85c8d6c1e46c83ea/packages/core/src/StateNode.ts#L211
2. each `StateNode` was reading there `context` of the machine: https://github.com/statelyai/xstate/blob/68b24d7c4ce7fa8920edab0a85c8d6c1e46c83ea/packages/core/src/StateNode.ts#L216
3. `StateMachine['context']` is a getter as well: https://github.com/statelyai/xstate/blob/68b24d7c4ce7fa8920edab0a85c8d6c1e46c83ea/packages/core/src/StateMachine.ts#L90-L92

This means that we were calling "lazy context" functions like crazy~